src/libraries/System.Linq/src/System/Linq/SegmentedArrayBuilder.cs

```diff
@@ -51,30 +52,44 @@ public void Dispose()
             int segmentsCount = _segmentsCount;
             if (segmentsCount != 0)
             {
-                ReadOnlySpan<T[]> segments = _segments;
+                ReturnArrays(segmentsCount);
+            }
+        }
 
-                // We need to return all rented arrays to the pool, and if the arrays contain any references,
-                // we want to clear them first so that the pool doesn't artificially root contained objects.
-                if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
-                {
-                    // Return all but the last segment. All of these are full and need to be entirely cleared.
-                    foreach (T[] segment in segments.Slice(0, segmentsCount - 1))
-                    {
-                        ArrayPool<T>.Shared.Return(segment, clearArray: true);
-                    }
+        private void ReturnArrays(int segmentsCount)
+        {
+            Debug.Assert(segmentsCount > 0);
+            ReadOnlySpan<T[]> segments = _segments;
 
-                    // For the last segment, we can clear only what we know was used.
-                    T[] currentSegment = segments[segmentsCount - 1];
-                    Array.Clear(currentSegment, 0, _countInCurrentSegment);
-                    ArrayPool<T>.Shared.Return(currentSegment);
+            // We need to return all rented arrays to the pool, and if the arrays contain any references,
+            // we want to clear them first so that the pool doesn't artificially root contained objects.
+            if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
+            {
+                // Return all but the last segment. All of these are full and need to be entirely cleared.
+                segmentsCount--;
+                foreach (T[] segment in segments.Slice(0, segmentsCount))
+                {
+                    Array.Clear(segment);
```

Comment:
What's the difference between clearing the segment here explicitly vs using the `clearArray` parameter?

Reply:
Guessing it's for consistency with the final segment?

Reply:
A combination of consistency and also not wanting to hurt branch prediction (for the `if (clearArray)` inside of Return).